backend/remote-state/oss/client.go

```diff
@@ -0,0 +1,248 @@
+package oss
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/aliyun/aliyun-oss-go-sdk/oss"
+	multierror "github.com/hashicorp/go-multierror"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+	"log"
+	"sync"
+)
+
+type RemoteClient struct {
+	ossClient            *oss.Client
+	bucketName           string
+	stateFile            string
+	lockFile             string
+	serverSideEncryption bool
+	acl                  string
+	doLock               bool
+	info                 *state.LockInfo
+	mu                   sync.Mutex
+}
+
+func (c *RemoteClient) Get() (payload *remote.Payload, err error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	buf, err := c.getObj(c.stateFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there was no data, then return nil
+	if buf == nil || len(buf.Bytes()) == 0 {
+		log.Printf("[DEBUG] State %s has no data.", c.stateFile)
+		return nil, nil
+	}
+	md5 := md5.Sum(buf.Bytes())
+
+	payload = &remote.Payload{
+		Data: buf.Bytes(),
+		MD5:  md5[:],
+	}
+	return payload, nil
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return c.putObj(c.stateFile, data)
+}
+
+func (c *RemoteClient) Delete() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return c.deleteObj(c.stateFile)
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.doLock {
+		return "", nil
+	}
+
+	bucket, err := c.ossClient.Bucket(c.bucketName)
+	if err != nil {
+		return "", fmt.Errorf("Error getting bucket: %#v", err)
+	}
+
+	infoJson, err := json.Marshal(info)
+	if err != nil {
+		return "", err
+	}
+
+	if info.ID == "" {
+		lockID, err := uuid.GenerateUUID()
+		if err != nil {
+			return "", err
+		}
+		info.ID = lockID
+	}
+
+	info.Path = c.lockFile
+	if exist, err := bucket.IsObjectExist(info.Path); err != nil {
+		return "", fmt.Errorf("Estimating object %s is exist got an error: %#v", info.Path, err)
+	} else if !exist {
+		if err := c.putObj(info.Path, infoJson); err != nil {
```

Comment:
Checking the existence then putting the lock object creates a race condition.
Is there some way to write the object and have it rejected the request if it exists? (this is one of the reasons the s3 backend has to delegate locking to DynamoDB). Basically, will `putObj` fail if the object already exists?

Reply:
I am sorry for that now there is no better way, like s3 DynamoDB, to reject the request . At present, the method `putObj` will overwrite the existing object, [see details](https://github.com/aliyun/aliyun-oss-go-sdk/blob/master/oss/bucket.go#L26). I think this question is not blocking issue for this PR, and this PR can be merged firstly. In the next, I will continue to find the better way like s3 DynamoDB to avoid some request.

Reply:
@jbardin Do you have other ideas?

Reply:
I don't think we want to advertise the storage as safe for concurrent use if the locking mechanism isn't actually safe. For now it's probably better to remove the locking altogether so there is no confusion about what this backend can provide.

Reply:
HI @jbardin Thanks for your suggestion. I want to make the locking mechanism safe by using alibaba cloud tablestore. But, when I tried to make a test using a custom terraform package, I happened a issue: `Error: Failed to instantiate provider "alicloud" to obtain schema: Incompatible API version with plugin. Plugin version: 4, Client versions: [5]`. Can you give me some help? Thanks in advance.

Reply:
@xiaozhu36, the provider needs to be updated to work with terraform 0.12 (the master branch).

Reply:
HI @jbardin Thanks for your feedback. I have updated but the error still exist.

Reply:
If the error says `Plugin version: 4`, that means the plugin you are executing was compiled with an old version of the terraform packages. The only way to fix that is to compile and execute the provider plugin using the latest version of the core package (which for now is `master`), and we suggest using go modules to do this.

Reply:
HI @jbardin I have add tablestore config to store the state lock. Please review it.